pincer/objects/message/message.py

```diff
@@ -1,79 +1,129 @@
+
+if TYPE_CHECKING:
+    from ..app import InteractionFlags
+    from .component import MessageComponent
+
+
+@dataclass
+class AllowedMentions(APIObject):
+    parse: List[AllowedMentionTypes]
+    roles: List[Union[Role, Snowflake]]
+    users: List[Union[User, Snowflake]]
+    reply: bool = True
+
+    def to_dict(self):
+        def get_str_id(obj: Union[Snowflake, User, Role]) -> str:
+            if hasattr(obj, "id"):
+                obj = obj.id
+
+            return str(obj)
+
+        return {
+            "parse": self.parse,
+            "roles": list(map(get_str_id, self.roles)),
+            "users": list(map(get_str_id, self.users)),
+            "replied_user": self.reply
+        }
+
+
+@dataclass
+class Message:
+    # TODO: Write docs
+    content: str = ''
+    attachments: Optional[File] = None
+    tts: Optional[bool] = False
+    embeds: Optional[List[Embed]] = None
+    allowed_mentions: Optional[AllowedMentions] = None
+    components: Optional[List[MessageComponent]] = None
+    flags: Optional[InteractionFlags] = None
+    type: Optional[CallbackType] = None
+
+    def __post_init__(self):
+
+        attch = []
+
+        for count,value in enumerate(self.attachments):
+            if isinstance(value,File):
+                attch += [value]
+            elif isinstance(value,Image):
+                attch += [File.from_image(
+                    value,
+                    f"file{count}.png",
+                )]
+            elif isinstance(value,str):
+                attch += [
+                    File.from_file(value)
+                ]
+            else:
+                raise ValueError(f"Attachment {count} is invalid type")
+
+        self.attachments = attch
+
+
+    def to_dict(self):
+        if len(self.content) < 1 and not self.embeds and not self.attachments:
+            raise CommandReturnIsEmpty("Cannot return empty message.")
+
+        allowed_mentions = (
+            self.allowed_mentions.to_dict()
+            if self.allowed_mentions else {}
+        )
+
+
+        # Attachments aren't serialized because they are not sent as part of the json
+        resp = {
+            "content": self.content,
+            "tts": self.tts,
+            "flags": self.flags,
+            "embeds": [embed.to_dict() for embed in (self.embeds or [])],
+            "allowed_mentions": allowed_mentions,
+            "components": [
+                components.to_dict() for components in (self.components or [])
+            ]
+        }
+
+        return {
+            "type": self.type or CallbackType.MESSAGE,
+            "data": {k: i for k, i in resp.items() if i}
+        }
+
+    def serialize(self) -> Tuple[str,Union[Payload,Dict]]:
+        """
+        Creates the data that the discord API wants for the message object
+
+        :return: (content_type, data)
+        """
+
+        if self.attachments:
+            form = FormData()
+            form.add_field("payload_json",json.dumps(self.to_dict()))
```

Comment:
i believe `form` could have chaining, and maybe a `.add_fields` method. This would allow us to write someone as 
```py
return (
    FormData()
        .add_field(...)
        .add_fields(self.attachments, lambda file: ("file",file.content, file.filename))
).headers["Content-Type"], payload

```
here.

Reply:
It doesn't look like it has chaining.
I can replace the for loop with add_fields() but Im not sure which is more readable.

```python
        form.add_fields(
            *(
                (file.filename, file.content)
                for file in self.attachments
            )
        )
```
```python
        for file in self.attachments:
            form.add_field("file", file.content, filename=file.filename)
```

Reply:
I mean that why i said `could` haha, yes unfortunately doesn't hav, but may be not to hard to add.
From your 2 proposition, the 2nd one sound more readable

Reply:
Im a little confused about what you want. I didn't write the class so Im at the mercy of the aiohttp gods.